src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,409 @@
+package k8sd
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/canonical/go-dqlite"
+	"github.com/canonical/go-dqlite/app"
+	"github.com/canonical/go-dqlite/client"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/termios"
+	"github.com/canonical/microcluster/v2/cluster"
+	"github.com/canonical/microcluster/v2/microcluster"
+	"github.com/spf13/cobra"
+	"golang.org/x/sys/unix"
+	"gopkg.in/yaml.v2"
+
+	"github.com/canonical/k8s/pkg/utils"
+)
+
+const recoveryConfirmation = `You should only run this command if:
+ - A quorum of cluster members is permanently lost
+ - You are *absolutely* sure all k8s daemons are stopped (sudo snap stop k8s)
+ - This instance has the most up to date database
+
+Do you want to proceed? (yes/no): `
+
+const clusterYamlRecoveryComment = `# Member roles can be modified. Unrecoverable nodes should be given the role "spare".
+#
+# "voter" - Voting member of the database. A majority of voters is a quorum.
+# "stand-by" - Non-voting member of the database; can be promoted to voter.
+# "spare" - Not a member of the database.
+#
+# The edit is aborted if:
+# - the number of members changes
+# - the name of any member changes
+# - the ID of any member changes
+# - the address of any member changes
+# - no changes are made
+`
+
+const infoYamlRecoveryComment = `# Verify the ID, address and role of the local node.
+#
+# Cluster members:
+`
+
+const daemonYamlRecoveryComment = `# Verify the name and address of the local node.
+#
+# Cluster members:
+`
+
+// Used as part of a regex search, avoid adding special characters.
+const yamlHelperCommentFooter = "# ------- everything below will be written -------"
+
+var clusterRecoverOpts struct {
+	K8sDqliteStateDir string
+}
+
+
+func newClusterRecoverCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "cluster-recover",
+		Short: "Recover the cluster from this member if quorum is lost",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := recoveryCmdPrechecks(); err != nil {
+				return err;
+			}
+
+			k8sdTarballPath, err := recoverK8sd();
+			if err != nil {
+				return fmt.Errorf("Failed to recover k8sd, error: %w.", err);
+			}
+			cmd.Printf("K8sd cluster changes applied; new database state saved to %s\n\n",
+					   k8sdTarballPath)
+			cmd.Printf("*Before* starting any cluster member, copy %s to %s " +
+					   "on all remaining cluster members.\n\n",
+					   k8sdTarballPath, k8sdTarballPath)
```

Comment:
ah, so microcluster will look up this specific path on startup and if a recovery archive exists on that path, it will load it?
I think then `/tmp` is not the best path for this. I'd recommend the same one as microcloud, e.g. `/var/snap/k8s/common/k8sd-recovery.tar.gz`

Reply:
Microcluster will load the k8sd archive on startup but not the k8s-dqlite database. That one needs to be handled manually.

We receive the state directories as CLI arguments and it probably wouldn't be safe to use the parent of the state dir. At the same time, I saw that microcluster checks if it runs as a snap, so we probably can't make that assumption either.

Reply:
For what is worth, k8sd/microcluster uses a subdir for the database files (``$statedir/database``) and stores the backups in the state dir. k8s-dqlite doesn't have a separate subdir for the database files, so storing a backup of $statedir in the state dir itself would not be ideal.

Reply:
Ok, I'll put the backups there.